core/pubnub_coreapi.c

```diff
@@ -297,6 +299,45 @@ enum pubnub_res pubnub_set_state(pubnub_t*   pb,
         pb->core.last_result = PNR_STARTED;
         pbnc_fsm(pb);
         rslt = pb->core.last_result;
+        if (rslt == PNR_STARTED){
+            int ch_cnt = 0;
+            int cg_cnt = 0;
+            char * json_state = (char*)malloc(strlen(state) + (channel ? strlen(channel) : 1) + (channel_group ? strlen(channel_group) : 1) + 20);
+            memcpy(json_state, "{", 1);
+            if (channel) {
+                char* chan_token = strtok((char*)channel, ",");
+                while( chan_token != NULL ) {
+                    if (0 != strncmp((char*)" ", chan_token, 1)) {
+                        if (ch_cnt > 0) { strcat(json_state, ","); }
+                        char* ch_state = (char*)malloc(strlen(state) + strlen(chan_token) + 3);
+                        sprintf(ch_state, "\"%s\":%s", chan_token, state);
+                        strcat(json_state, ch_state);
+                        ch_cnt++;
+                        free(ch_state);
+                    }
+                    chan_token = strtok(NULL, ",");
+                }
+            }
+            if (channel_group) {
+                char* cg_token = strtok((char*)channel_group, ",");
+                while( cg_token != NULL ) {
+                    if (0 != strncmp((char*)" ", cg_token, 1)) {
+                        if (cg_cnt > 0 || ch_cnt > 0) { strcat(json_state, ","); }
+                        char* cg_state = (char*)malloc(strlen(state) + strlen(cg_token) + 3);
+                        sprintf(cg_state, "\"%s\":%s", cg_token, state);
+                        strcat(json_state, cg_state);
+                        cg_cnt++;
+                        free(cg_state);
+                    }
+                    cg_token = strtok(NULL, ",");
+                }
+            }
+            strcat(json_state, "}");
+            PUBNUB_LOG_DEBUG("formatted state is %s\n", json_state);
+
+            pb->core.state = json_state;
+            free(json_state);
```

Comment:
Maybe it's my lack of `C` knowledge, but it looks like it's going to set `pb->core.state` to point at the same memory as `json_state` but right away after that it's freeing this memory. Am I wrong with this?

Reply:
I am still testing. Lets hold our horses.